compiler/resolution/callDestructors.cpp

```diff
@@ -438,8 +542,44 @@ void ReturnByRef::transformMove(CallExpr* moveExpr)
 
   Expr*     lhs      = moveExpr->get(1);
   CallExpr* callExpr = toCallExpr(moveExpr->get(2));
+  CallExpr* copyExpr = NULL;
   Symbol*   useLhs   = toSymExpr(lhs)->var;
   Symbol*   refVar   = newTemp("ret_to_arg_ref_tmp_", useLhs->type->refType);
+  FnSymbol* fn       = callExpr->isResolved();
+
+  if (fn == NULL)
```

Comment:
That is what is going on.  This case was particularly challenged by my urge to write

else if (Expr* next = …) {

}


in the final clause since that pattern doesn’t combine with && AFAIK.  It does look a bit
odd at first sight though doesn’t it?  I can see two obvious alternatives e.g

if (fn != NULL && fn->hasFlag(FLAG_CONSTRUCTOR) == false)
{
  if (Expr* next = moveExpr->next)
  {

  }
}


or

Expr* next = moveExpr->next;

if (fn != NULL && fn->hasFlag(..)  && next != NULL)
{

}


I am open to your thoughts on this.

Mike





On Dec 1, 2015, at 4:32 PM, Kyle Brady <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/callDestructors.cpp<https://github.com/chapel-lang/chapel/pull/2956#discussion_r46362281>:

>    Symbol*   useLhs   = toSymExpr(lhs)->var;
>    Symbol*   refVar   = newTemp("ret_to_arg_ref_tmp_", useLhs->type->refType);
> +  FnSymbol* fn       = callExpr->isResolved();
> +
> +  if (fn == NULL)


Are these purposefully blank to avoid &&s in the else if below?

—
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/2956/files#r46362281>.